libs/gi/ui/src/components/character/CharacterMultiSelectionModal.tsx

```diff
@@ -0,0 +1,500 @@
+  catTotal,
+  filterFunction,
+  sortFunction,
+} from '@genshin-optimizer/common/util'
+import { characterAsset } from '@genshin-optimizer/gi/assets'
+import type { CharacterKey } from '@genshin-optimizer/gi/consts'
+import {
+  allCharacterKeys,
+  allElementKeys,
+  allWeaponTypeKeys,
+} from '@genshin-optimizer/gi/consts'
+import type { ICachedCharacter } from '@genshin-optimizer/gi/db'
+import {
+  useCharMeta,
+  useCharacter,
+  useDBMeta,
+  useDatabase,
+} from '@genshin-optimizer/gi/db-ui'
+import type { CharacterSheet } from '@genshin-optimizer/gi/sheets'
+import { getCharSheet } from '@genshin-optimizer/gi/sheets'
+import { getCharEle, getCharStat } from '@genshin-optimizer/gi/stats'
+import { ascensionMaxLevel } from '@genshin-optimizer/gi/util'
+import CloseIcon from '@mui/icons-material/Close'
+import FavoriteIcon from '@mui/icons-material/Favorite'
+import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder'
+import type { TooltipProps } from '@mui/material'
+import {
+  Box,
+  CardActionArea,
+  CardContent,
+  Divider,
+  Grid,
+  IconButton,
+  TextField,
+  Tooltip,
+  Typography,
+  styled,
+  tooltipClasses,
+} from '@mui/material'
+import type { ChangeEvent } from 'react'
+import {
+  useContext,
+  useDeferredValue,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react'
+import { useTranslation } from 'react-i18next'
+import { DataContext, SillyContext } from '../../context'
+import { iconAsset } from '../../util/iconAsset'
+import { ElementToggle, WeaponToggle } from '../toggles'
+import { CharacterCard } from './CharacterCard'
+import type { CharacterSortKey } from './CharacterSort'
+import {
+  characterFilterConfigs,
+  characterSortConfigs,
+  characterSortMap,
+} from './CharacterSort'
+import { CharacterName } from './Trans'
+
+type characterFilter = (
+  characterKey: CharacterKey,
+  character: ICachedCharacter | undefined,
+  sheet: CharacterSheet
+) => boolean
+
+type CharacterMultiSelectionModalProps = {
+  show: boolean
+  newFirst?: boolean
+  onHide: () => void
+  onMultiSelect?: (cKeys: (CharacterKey | '')[]) => void
+  teamId: string
+  filter?: characterFilter
+}
+const sortKeys = Object.keys(characterSortMap)
+export function CharacterMultiSelectionModal({
+  show,
+  onHide,
+  onMultiSelect,
+  teamId,
+  filter = () => true,
+  newFirst = false,
+}: CharacterMultiSelectionModalProps) {
+  const { t } = useTranslation([
+    'page_character',
+    // Always load these 2 so character names are loaded for searching/sorting
+    'sillyWisher_charNames',
+    'charNames_gen',
+  ])
+  const { silly } = useContext(SillyContext)
+  const database = useDatabase()
+  const [state, setState] = useState(() => database.displayCharacter.get())
+  useEffect(
+    () => database.displayCharacter.follow((r, s) => setState(s)),
+    [database, setState]
+  )
+
+  const team = database.teams.get(teamId)!
+  const { loadoutData } = team
+  const currentTeamCharKeys : (CharacterKey | '')[] = useMemo(() => {
```

Comment:
This is the main reason that the child components are not updating:
![image](https://github.com/user-attachments/assets/bad198e2-8588-4307-a480-7be3fe5e0bee)
 because this object always has the same reference across renders, which does not trigger rerender in the child components. 
 
 Overall, because people will toggle the characters very fast, if you wait for `loadoutData` to come back, it would requires the whole 4 characters in the team to be recalced, which is very expensive/laggy.
 
 My recommendation would be to store the characters locally in a `const [teamCharKeys, setTeamCharKeys] = useState([])` and only push the change to the selection when you close the modal (I know you are doing half of this already)

Reply:
This is sort of how I would write this (very roughly)
```Typescript
const [teamCharKeys, setTeamCharKeys] = useState([null,null,null,null])
const { loadoutData } = useTeam(teamId)
// update teamCharKeys when loadoutData changes
useEffect(()=>setTEamCharKeys(loadoutData.map((loadoutDatum) =>
	database.teamChars.get(loadoutDatum?.teamCharId)?.key ?? null
)),[setTeamCharKeys, loadoutData])
```